#Analyzing eBay Car Sales Data with Python

The goal of this project is to analyze 2016 data on used cars from eBay Kleinanzeigen, a section of the German eBay website. We'll be doing that using the Python libraries pandas and NumPy.

##Data cleaning

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We notice the 20-column dataframe has a lot of values written in German (for example, PS means Pferdestrke, which is horsepower). We also see that the dateCreated column seems redundant since we also have the dateCreawled column giving us the information. Interesting values to analyze would be, among others, powerPS, price, yearOfResgistration and odometer.

We can also observe that some columns have null values, and that the column names are in camelcase instead of snakecase which is the formality in Python.

In [5]:
autos.describe(include="all")

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

We notice that the seller and offerType column almost have the same value for each row: they won't give us much information and we'll drop them.

Two of the interesting columns that we identified earlier (price  and odometer) are stored as strings, but we'd like them to be stored as numeric values instead. The next thing we'll do is convert them to numeric values.

In [6]:
autos_price = autos["price"]
autos_price_clean = []
for p in autos_price:
    p = str(p)
    p = p.replace('$', '')
    p = p.replace(',', '')
    autos_price_clean.append(p)

autos["price"] = autos_price_clean   

autos["price"] = autos["price"].astype(np.float64)
    
autos_odometer = autos["odometer"]
autos_odometer_clean = []
for o in autos_odometer:
    o = str(o)
    o = o.replace("km", "")
    o= o.replace(',', '')
    autos_odometer_clean.append(o)
    
autos["odometer"] = autos_odometer_clean
    
autos["odometer"] = autos["odometer"].astype(np.int64)

autos = autos.rename(columns = {"odometer" :"odometer_km"})

autos.describe(include="all")

dateCrawled         name  seller offerType         price  \
count                 50000        50000   50000     50000  5.000000e+04   
unique                48213        38754       2         2           NaN   
top     2016-03-27 22:55:05  Ford_Fiesta  privat   Angebot           NaN   
freq                      3           78   49999     49999           NaN   
mean                    NaN          NaN     NaN       NaN  9.840044e+03   
std                     NaN          NaN     NaN       NaN  4.811044e+05   
min                     NaN          NaN     NaN       NaN  0.000000e+00   
25%                     NaN          NaN     NaN       NaN  1.100000e+03   
50%                     NaN          NaN     NaN       NaN  2.950000e+03   
75%                     NaN          NaN     NaN       NaN  7.200000e+03   
max                     NaN          NaN     NaN       NaN  1.000000e+08   

       abtest vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count   50000       44905        50000.000000    47320  50000.000000  47242   
unique      2           8                 NaN        2           NaN    245   
top      test   limousine                 NaN  manuell           NaN   golf   
freq    25756       12859                 NaN    36993           NaN   4024   
mean      NaN         NaN         2005.073280      NaN    116.355920    NaN   
std       NaN         NaN          105.712813      NaN    209.216627    NaN   
min       NaN         NaN         1000.000000      NaN      0.000000    NaN   
25%       NaN         NaN         1999.000000      NaN     70.000000    NaN   
50%       NaN         NaN         2003.000000      NaN    105.000000    NaN   
75%       NaN         NaN         2008.000000      NaN    150.000000    NaN   
max       NaN         NaN         9999.000000      NaN  17700.000000    NaN   

          odometer_km  monthOfRegistration fuelType       brand  \
count    50000.000000         50000.000000    45518       50000   
unique            NaN                  NaN        7          40   
top               NaN                  NaN   benzin  volkswagen   
freq              NaN                  NaN    30107       10687   
mean    125732.700000             5.723360      NaN         NaN   
std      40042.211706             3.711984      NaN         NaN   
min       5000.000000             0.000000      NaN         NaN   
25%     125000.000000             3.000000      NaN         NaN   
50%     150000.000000             6.000000      NaN         NaN   
75%     150000.000000             9.000000      NaN         NaN   
max     150000.000000            12.000000      NaN         NaN   

       notRepairedDamage          dateCreated  nrOfPictures    postalCode  \
count              40171                50000       50000.0  50000.000000   
unique                 2                   76           NaN           NaN   
top                 nein  2016-04-03 00:00:00           NaN           NaN   
freq               35232                 1946           NaN           NaN   
mean                 NaN                  NaN           0.0  50813.627300   
std                  NaN                  NaN           0.0  25779.747957   
min                  NaN                  NaN           0.0   1067.000000   
25%                  NaN                  NaN           0.0  30451.000000   
50%                  NaN                  NaN           0.0  49577.000000   
75%                  NaN                  NaN           0.0  71540.000000   
max                  NaN                  NaN           0.0  99998.000000   

                   lastSeen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

While we're at it, let's rename all the columns in snakecase format.

In [7]:
autos = autos.rename(columns = {"dateCrawled" : "date_crawled", "offerType": "offer_type", "vehicleType": "vehicle_type", "yearOfRegistration": "year_of_registration", "powerPS": "power_ps", "monthOfRegistration": "month_of_registration", "fuelType": "fuel_type", "notRepairedDamage": "not_repaired_damage", "dateCreated": "date_created", "nrOfPictures": "nr_of_pictures", "postalCode": "postal_code", "lastSeen": "last_seen" })

We will now analyse the price and odometer_km columns to see if there are unrealistic values.

Let's start with the price column.

In [8]:
autos_price = autos["price"]

autos_price.describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [9]:
autos_price_values = autos_price.value_counts()
print(autos_price_values)

0.0           1421
500.0          781
1500.0         734
2500.0         643
1200.0         639
1000.0         639
600.0          531
800.0          498
3500.0         498
2000.0         460
999.0          434
750.0          433
900.0          420
650.0          419
850.0          410
700.0          395
4500.0         394
300.0          384
2200.0         382
950.0          379
1100.0         376
1300.0         371
3000.0         365
550.0          356
1800.0         355
5500.0         340
1250.0         335
350.0          335
1600.0         327
1999.0         322
              ... 
2225.0           1
69997.0          1
139997.0         1
69999.0          1
4780.0           1
8930.0           1
21599.0          1
15911.0          1
10000000.0       1
5180.0           1
919.0            1
1247.0           1
5998.0           1
27020.0          1
21888.0          1
46500.0          1
2001.0           1
2459.0           1
345000.0         1
34940.0          1
2785.0           1
5248.0      

In [10]:
autos_price_values.sort_index(ascending = "False")

0.0           1421
1.0            156
2.0              3
3.0              1
5.0              2
8.0              1
9.0              1
10.0             7
11.0             2
12.0             3
13.0             2
14.0             1
15.0             2
17.0             3
18.0             1
20.0             4
25.0             5
29.0             1
30.0             7
35.0             1
40.0             6
45.0             4
47.0             1
49.0             4
50.0            49
55.0             2
59.0             1
60.0             9
65.0             5
66.0             1
              ... 
151990.0         1
155000.0         1
163500.0         1
163991.0         1
169000.0         1
169999.0         1
175000.0         1
180000.0         1
190000.0         1
194000.0         1
197000.0         1
198000.0         1
220000.0         1
250000.0         1
259000.0         1
265000.0         1
295000.0         1
299000.0         1
345000.0         1
350000.0         1
999990.0         1
999999.0    

We see that there are a lot of price values under 100$, and also some over a million $. Because we are interested in the cars that have a certain quality and a realistic and quite affordable price, we will only keep cars that have a price between 200$ and 400000$.

In [11]:
autos = autos[autos["price"].between(200, 400000)]
autos["price"].describe()

count     47645.000000
mean       6000.707273
std        9110.783444
min         200.000000
25%        1300.000000
50%        3190.000000
75%        7500.000000
max      350000.000000
Name: price, dtype: float64

We can see that our filter only removed around 2000 cars from our 50000-car dataset, which is proof that our filter only discarded extreme values. The average price of a car in our set is now a reasonable 6000$.

Now let's analyze the odometer_km column.

In [12]:
autos_odometer_km = autos["odometer_km"]
autos_odometer_km.describe()

count     47645.000000
mean     125887.501312
std       39482.911790
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [13]:
autos_odometer_km_values = autos_odometer_km.value_counts(dropna = False)
print(autos_odometer_km_values)

150000    30781
125000     4998
100000     2083
90000      1720
80000      1407
70000      1209
60000      1145
50000      1007
40000       813
30000       773
20000       747
5000        720
10000       242
Name: odometer_km, dtype: int64


In [14]:
autos_odometer_km_values.sort_index(ascending = "False")

5000        720
10000       242
20000       747
30000       773
40000       813
50000      1007
60000      1145
70000      1209
80000      1407
90000      1720
100000     2083
125000     4998
150000    30781
Name: odometer_km, dtype: int64

These odometer values all make sense and don't have null values: we won't discard any row based on this column.

Now let's analyze the columns that are date values(some are added by the crawler and some by the website itself):
- date_crawled: added by the crawler, a string
- date_created: from the website, a string
- last_seen: added by the crawler, a string
- registration_month: from the website
- registration_year: from the website

We'll start by examining the three string columns.

In [15]:
autos[['date_crawled','date_created','last_seen']][0:5]

date_crawled         date_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The exact time isn't pertinent to us: we'll only keep the date from these values.

In [16]:
autos["date_crawled"] = autos["date_crawled"].str[:10]
autos["date_created"] = autos["date_created"].str[:10]
autos["last_seen"] = autos["last_seen"].str[:10]

autos[['date_crawled','date_created','last_seen']][0:5]

date_crawled date_created   last_seen
0   2016-03-26   2016-03-26  2016-04-06
1   2016-04-04   2016-04-04  2016-04-06
2   2016-03-26   2016-03-26  2016-04-06
3   2016-03-12   2016-03-12  2016-03-15
4   2016-04-01   2016-04-01  2016-04-01

Now let's verify if date_crawled and date_created are really identical.

In [17]:
autos[['date_crawled','date_created']]

date_crawled date_created
0       2016-03-26   2016-03-26
1       2016-04-04   2016-04-04
2       2016-03-26   2016-03-26
3       2016-03-12   2016-03-12
4       2016-04-01   2016-04-01
5       2016-03-21   2016-03-21
6       2016-03-20   2016-03-20
7       2016-03-16   2016-03-16
8       2016-03-22   2016-03-22
9       2016-03-16   2016-03-16
10      2016-03-15   2016-03-14
11      2016-03-16   2016-03-16
12      2016-03-31   2016-03-31
13      2016-03-23   2016-03-23
14      2016-03-23   2016-03-23
15      2016-04-01   2016-04-01
16      2016-03-16   2016-03-16
17      2016-03-29   2016-03-29
18      2016-03-26   2016-03-26
19      2016-03-17   2016-03-17
20      2016-03-05   2016-03-05
21      2016-03-06   2016-03-06
22      2016-03-28   2016-03-28
23      2016-03-10   2016-03-10
24      2016-04-03   2016-04-03
26      2016-04-03   2016-04-03
28      2016-03-19   2016-03-19
29      2016-04-02   2016-04-02
31      2016-03-14   2016-03-14
32      2016-03-20   2016-03-20
...            ...          ...
49967   2016-03-12   2016-03-12
49968   2016-04-01   2016-04-01
49969   2016-03-17   2016-03-17
49970   2016-03-21   2016-03-21
49971   2016-03-29   2016-03-29
49972   2016-03-26   2016-03-26
49973   2016-03-27   2016-03-27
49975   2016-03-27   2016-03-27
49976   2016-03-19   2016-03-19
49977   2016-03-31   2016-03-31
49978   2016-04-04   2016-04-04
49979   2016-03-20   2016-03-20
49980   2016-03-12   2016-03-12
49981   2016-03-15   2016-03-15
49982   2016-03-29   2016-03-29
49983   2016-03-06   2016-03-06
49985   2016-04-02   2016-04-02
49986   2016-04-04   2016-04-04
49987   2016-03-22   2016-03-22
49988   2016-03-28   2016-03-28
49990   2016-03-21   2016-03-21
49991   2016-03-06   2016-03-06
49992   2016-03-10   2016-03-10
49993   2016-03-15   2016-03-15
49994   2016-03-22   2016-03-22
49995   2016-03-27   2016-03-27
49996   2016-03-28   2016-03-28
49997   2016-04-02   2016-04-02
49998   2016-03-08   2016-03-08
49999   2016-03-14   2016-03-13

[47645 rows x 2 columns]

The values are almost all indentical. If they aren't, the dates are close and they don't give much information. We will drop the date_crawled column.

In [18]:
autos = autos.drop(columns = "date_crawled")

Now we'll analyze the distribution of the dates of the ad creation.

In [19]:
autos_date_created = autos["date_created"].value_counts(normalize=True, dropna=False)
print(autos_date_created)

2016-04-03    0.038976
2016-03-20    0.037884
2016-03-21    0.037570
2016-04-04    0.036919
2016-03-12    0.036688
2016-04-02    0.035366
2016-03-14    0.035177
2016-03-28    0.035114
2016-03-07    0.034715
2016-03-15    0.033959
2016-03-29    0.033959
2016-04-01    0.033792
2016-03-30    0.033708
2016-03-19    0.033519
2016-03-09    0.033141
2016-03-08    0.033120
2016-03-11    0.033015
2016-03-22    0.032532
2016-03-26    0.032259
2016-03-23    0.032217
2016-03-10    0.032029
2016-03-31    0.031882
2016-03-25    0.031546
2016-03-17    0.031210
2016-03-27    0.031147
2016-03-16    0.029951
2016-03-24    0.029258
2016-03-05    0.022940
2016-03-13    0.017106
2016-03-06    0.015322
                ...   
2016-02-24    0.000042
2016-01-10    0.000042
2016-02-26    0.000042
2016-02-12    0.000042
2016-02-18    0.000042
2016-02-05    0.000042
2016-02-14    0.000042
2016-02-02    0.000042
2016-02-20    0.000042
2015-06-11    0.000021
2016-01-29    0.000021
2016-02-11    0.000021
2016-01-22 

In [20]:
autos_date_created.sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000042
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000063
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000042
2016-02-05    0.000042
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000042
2016-02-14    0.000042
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000042
2016-02-19    0.000063
2016-02-20    0.000042
2016-02-21    0.000063
                ...   
2016-03-09    0.033141
2016-03-10    0.032029
2016-03-11    0.033015
2016-03-12    0.036688
2016-03-13    0.017106
2016-03-14    0.035177
2016-03-15    0.033959
2016-03-16    0.029951
2016-03-17    0.031210
2016-03-18    0.013454
2016-03-19    0.033519
2016-03-20    0.037884
2016-03-21 

The dates of the ad creations make sense, as they are all between 2015-2016 and there aren't any null values.

Now let's do the same with the last_seen column.

In [21]:
autos_last_seen = autos["last_seen"].value_counts(normalize=True, dropna=False)
print(autos_last_seen)

2016-04-06    0.222437
2016-04-07    0.132669
2016-04-05    0.125470
2016-03-17    0.028083
2016-04-03    0.025102
2016-04-02    0.024808
2016-04-04    0.024620
2016-03-30    0.024578
2016-03-12    0.023906
2016-03-31    0.023864
2016-04-01    0.022899
2016-03-29    0.022185
2016-03-22    0.021408
2016-03-28    0.020758
2016-03-20    0.020653
2016-03-21    0.020506
2016-03-24    0.019624
2016-03-25    0.019079
2016-03-23    0.018491
2016-03-26    0.016686
2016-03-16    0.016287
2016-03-15    0.015720
2016-03-19    0.015615
2016-03-27    0.015511
2016-03-14    0.012530
2016-03-11    0.012299
2016-03-10    0.010473
2016-03-09    0.009613
2016-03-13    0.008899
2016-03-18    0.007283
2016-03-08    0.007178
2016-03-07    0.005373
2016-03-06    0.004303
2016-03-05    0.001091
Name: last_seen, dtype: float64


In [22]:
autos_last_seen.sort_index()

2016-03-05    0.001091
2016-03-06    0.004303
2016-03-07    0.005373
2016-03-08    0.007178
2016-03-09    0.009613
2016-03-10    0.010473
2016-03-11    0.012299
2016-03-12    0.023906
2016-03-13    0.008899
2016-03-14    0.012530
2016-03-15    0.015720
2016-03-16    0.016287
2016-03-17    0.028083
2016-03-18    0.007283
2016-03-19    0.015615
2016-03-20    0.020653
2016-03-21    0.020506
2016-03-22    0.021408
2016-03-23    0.018491
2016-03-24    0.019624
2016-03-25    0.019079
2016-03-26    0.016686
2016-03-27    0.015511
2016-03-28    0.020758
2016-03-29    0.022185
2016-03-30    0.024578
2016-03-31    0.023864
2016-04-01    0.022899
2016-04-02    0.024808
2016-04-03    0.025102
2016-04-04    0.024620
2016-04-05    0.125470
2016-04-06    0.222437
2016-04-07    0.132669
Name: last_seen, dtype: float64

Again, these dates make sense and there's a pertinent value for each row. We won't remove any row based on date_created and last_seen columns.

We can now look at the registration years of the cars.

In [23]:
autos["year_of_registration"].describe()

count    47645.000000
mean      2004.800084
std         88.423872
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

We see that there is a year value of 1000 and another of 9999. Knowing that this collection of data was made in 2016, and that cars only exist since the early 1900s ,we will only keep the cars that have a registration year in this range.

In [24]:
autos = autos[autos["year_of_registration"].between(1900, 2016)]
autos["year_of_registration"].describe()

count    45786.000000
mean      2002.993098
std          7.113188
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: year_of_registration, dtype: float64

In [25]:
autos["year_of_registration"].value_counts(normalize=True)

2000    0.065937
2005    0.063469
1999    0.062137
2004    0.058839
2003    0.058795
2006    0.058249
2001    0.057026
2002    0.053947
1998    0.049906
2007    0.049622
2008    0.048181
2009    0.045407
1997    0.040449
2011    0.035316
2010    0.034661
2012    0.028546
1996    0.028437
2016    0.025357
1995    0.024440
2013    0.017407
2014    0.014328
1994    0.012952
1993    0.008867
2015    0.008212
1992    0.007753
1991    0.007207
1990    0.006989
1989    0.003691
1988    0.002905
1985    0.001922
          ...   
1976    0.000459
1960    0.000459
1969    0.000415
1975    0.000393
1965    0.000371
1964    0.000262
1963    0.000175
1961    0.000131
1959    0.000131
1962    0.000087
1956    0.000087
1937    0.000087
1958    0.000087
1955    0.000044
1954    0.000044
1957    0.000044
1951    0.000044
1910    0.000044
1934    0.000044
1941    0.000044
1927    0.000022
1929    0.000022
1950    0.000022
1931    0.000022
1948    0.000022
1938    0.000022
1939    0.000022
1953    0.0000

The oldest car in our new set was registered in 1910 and the newest in 2016, which is coherent. We can also see that the most popular registration years for the cars are in the decade 2000-2009

One last column from our set that we'd like to clean is power_ps (horsepower). Let's see its description.

In [26]:
autos["power_ps"].describe()

count    45786.000000
mean       119.168654
std        185.462987
min          0.000000
25%         75.000000
50%        109.000000
75%        150.000000
max      17700.000000
Name: power_ps, dtype: float64

HP over 1700HP makes no sense because the most powerful car in 2020 has around 1600HP, and we see that we have a maximum value of 17700 HP. Also, we'll decide to keep only car with 60HP for our analysis, since cars who don't have this power aren't very useful in today's world (except for collectors).

In [27]:
autos = autos[autos["power_ps"].between(60,1700)]
autos["power_ps"].describe()

count    38894.000000
mean       133.286111
std         66.463419
min         60.000000
25%         90.000000
50%        120.000000
75%        160.000000
max       1405.000000
Name: power_ps, dtype: float64

These values make a lot more sense, as we are left with around 39000 cars with a valid horsepower value.

##Data analysis and metrics

Now that we have a valid and workable dataset, let's find out some pertinent metrics and information.

First let's calculate an interesting value: the mean price per brand. We will do that for the top 25 brands by popularity in our set.

In [28]:
#we create a dictionary which will store our mean prices per brand
top25_brands_mean_price = {}

#We select top 20 most popular car brands
top25_brands = autos["brand"].value_counts().head(25)
print("Most popular brands:")
print(top25_brands)

#iteration of the brands and calculation of the mean price
for b in top25_brands.index:
    brand_prices = autos.loc[autos["brand"]== b, "price"]
    mean_brand_price = brand_prices.mean()
    #adding the data to the dictionary
    top25_brands_mean_price[b] = mean_brand_price
    
print("Mean  price per brand:")
print(top25_brands_mean_price)
    

Most popular brands:
volkswagen        8035
bmw               4780
mercedes_benz     4048
opel              3844
audi              3756
ford              2694
renault           1383
peugeot           1193
fiat               851
skoda              695
seat               688
mazda              629
citroen            556
toyota             543
nissan             542
hyundai            412
mini               391
volvo              390
smart              363
mitsubishi         337
honda              336
kia                297
sonstige_autos     284
alfa_romeo         282
porsche            264
Name: brand, dtype: int64
Mean  price per brand:
{'alfa_romeo': 4198.563829787234, 'kia': 6299.272727272727, 'porsche': 48539.318181818184, 'citroen': 3985.6978417266187, 'opel': 3520.2788761706556, 'nissan': 5731.649446494465, 'seat': 5189.594476744186, 'hyundai': 5828.111650485437, 'renault': 3215.6305133767173, 'sonstige_autos': 15518.274647887323, 'mercedes_benz': 9001.05459486166, 'mazda': 4450.3

Here we see that the most popular brands are Volkswagen, BMW and Opel. We also printed the mean prices per brands, but we'll see better if we order them. Let's print this result in order of mean price:

In [29]:
print(sorted(top25_brands_mean_price.items(),key = 
             lambda kv:(kv[1], kv[0])))

[('renault', 3215.6305133767173), ('fiat', 3342.639247943596), ('peugeot', 3383.4534786253143), ('opel', 3520.2788761706556), ('mitsubishi', 3611.433234421365), ('citroen', 3985.6978417266187), ('ford', 4143.342242019302), ('alfa_romeo', 4198.563829787234), ('honda', 4291.839285714285), ('mazda', 4450.344992050875), ('smart', 4509.0), ('seat', 5189.594476744186), ('volvo', 5211.882051282051), ('toyota', 5296.141804788213), ('nissan', 5731.649446494465), ('hyundai', 5828.111650485437), ('volkswagen', 6136.304542626011), ('kia', 6299.272727272727), ('skoda', 6837.407194244604), ('bmw', 8627.34769874477), ('mercedes_benz', 9001.05459486166), ('audi', 9748.255324813632), ('mini', 10746.565217391304), ('sonstige_autos', 15518.274647887323), ('porsche', 48539.318181818184)]


When observing this result, we can see that Renault, Fiat and Peugeot are the cheapest used car brands in this market. On the opposite side, Porsche, Sonstige Autos and Mini are the most expensive cars in our dataset among the top 25 brands by popularity. It's interesting to see that Volkswagen, the most popular brand in our set is located around the middle of the pack, while BMW and Mercedes-Benz, the second and third most popular brands, are the 6th and 5th most expensive brand in our top 25 list, respectively.

Another very interesting statistic to extract from our dataset for people looking for a bargain on a performance car is the horsepower/price ratio.

Let's create the list of the top 10 individual cars in the dataset based on this ratio.

In [49]:
#calculate the ratio and add it as a new column
autos["ps_per_price"] = autos["power_ps"] / autos["price"]

#print information on the top 10 cars based on this ratio
top_10_ps_per_price = autos[["name", "brand", "model", "year_of_registration", "odometer_km", "price", "power_ps", "ps_per_price", "not_repaired_damage"]].sort_values(by = ["ps_per_price"], ascending=False).head(10)
top_10_ps_per_price

name           brand  \
19352                                      BMW_compact             bmw   
9405                                        VW_POLO_6N      volkswagen   
37651                         verkauf__bastler__export            opel   
21402                             Verkaufe_Ford_Fiesta            ford   
13573       A_140__Bastler_fahrzeug__mit_TÜV_bis_08/17   mercedes_benz   
11893  Original__Lamborgini_Diablo_Schutzhaube_91_2002  sonstige_autos   
2220                               Ford_ka_top_zustand            ford   
8376              e320_Unfall_Export_Motor_Getriebe_ok   mercedes_benz   
32805               W_lupo_um_sonderprais_zum_werkafen      volkswagen   
18268           Opel_Corsa_1_4_automatik_TÜV__10/2016_            opel   

          model  year_of_registration  odometer_km  price  power_ps  \
19352       3er                  1994       150000  399.0      1103   
9405       polo                  1995       150000  250.0       515   
37651     astra                  1999       150000  500.0      1001   
21402    fiesta                  1999       150000  400.0       754   
13573  a_klasse                  1999       150000  750.0      1400   
11893       NaN                  2002         5000  300.0       551   
2220         ka                  2005         5000  850.0      1003   
8376   e_klasse                  2000       150000  200.0       222   
32805      lupo                  1999        20000  990.0      1000   
18268     corsa                  1995       125000  500.0       500   

       ps_per_price not_repaired_damage  
19352      2.764411                 NaN  
9405       2.060000                 NaN  
37651      2.002000                 NaN  
21402      1.885000                 NaN  
13573      1.866667                 NaN  
11893      1.836667                 NaN  
2220       1.180000                nein  
8376       1.110000                  ja  
32805      1.010101                 NaN  
18268      1.000000                  ja

With this list, we can see that the car on which you'd pay the least per HP is a 1994 BMW Series 3 Compact. All of these cars are from the 90s or early 2000s and are on sale for under 1000$, which is a bargain if you're looking for performance. Notice that two of those cars have non-repaired damage.

In this project, we cleaned data from the German eBay site of used cars to make it pertinent and valid. We then used it to find the average price you would pay for a given car of the 25 most popular brands, and we also extracted a list of the best bargains on the site if you are a fan of performance by calculating the best horsepower / price ratios.